### PR TITLE
[ML] Serverless Endpoints

### PR DESCRIPTION
**Pull request by needuv**
# Description

Please add an informative description that covers that changes made by the pull request and link all relevant issues.

If an SDK is being regenerated based on a new swagger spec, a link to the pull request containing these swagger spec changes has been included above.

# All SDK Contribution checklist:
- [x] **The pull request does not introduce [breaking changes]**
- [x] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-python/blob/main/CONTRIBUTING.md).**

## General Guidelines and Best Practices
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-python/blob/main/CONTRIBUTING.md##building-and-testing)
- [x] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: sdk/ml/azure-ai-ml/azure/ai/ml/entities/_autogen_entities/models/_models.py [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+
+from .. import _model_base
+from .._model_base import rest_field
+
+if TYPE_CHECKING:
+    # pylint: disable=unused-import,ungrouped-imports
+    from .. import models as _models
+
+
+class AzureOpenAIDeployment(_model_base.Model):
```
Comment (annatisch) [inline sdk/ml/azure-ai-ml/azure/ai/ml/entities/_autogen_entities/models/_models.py:20]: All of these models will need to be added to `entities.__init__.py` to be made public.

File: sdk/ml/azure-ai-ml/azure/ai/ml/entities/_autogen_entities/models/_patch.py [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+                    return attr_value._validate()
+                rest_field_type = _get_rest_field_type(rest_field)
+                if attr_type != rest_field_type:
+                    raise ValueError(f"Type of attr {attr} is of type {attr_type}, not {rest_field_type}")
+
+
+@experimental
+class ServerlessEndpoint(_ServerlessEndpoint, ValidationMixin):
+    def _to_rest_object(self) -> RestServerlessEndpoint:
+        return RestServerlessEndpoint(
```
Comment (annatisch) [inline sdk/ml/azure-ai-ml/azure/ai/ml/entities/_autogen_entities/models/_patch.py:83]: These to-and-from methods shouldn't really be necessary, as the new-style generated models should already be in a REST-ready format.

File: sdk/ml/azure-ai-ml/azure/ai/ml/entities/_endpoint/serverless_endpoint.py [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+    ServerlessEndpointProperties as RestServerlessEndpointProperties,
+    ModelSettings as RestModelSettings,
+    Sku as RestSku,
+)
+
+from .endpoint import Endpoint
+
+
+@experimental
+class ServerlessEndpoint(Endpoint):
```
Comment (annatisch) [inline sdk/ml/azure-ai-ml/azure/ai/ml/entities/_endpoint/serverless_endpoint.py:23]: This model seems to be a duplicate of the generated-and-patched one in `_autogen_entities`?

Comment (needuv) [reply sdk/ml/azure-ai-ml/azure/ai/ml/entities/_endpoint/serverless_endpoint.py:23]: yes, thanks for pointing that out. I'll remove it

File: sdk/ml/azure-ai-ml/azure/ai/ml/entities/_autogen_entities/models/_patch.py (hunk 2 above)
Comment (needuv) [reply sdk/ml/azure-ai-ml/azure/ai/ml/entities/_autogen_entities/models/_patch.py:83]: What I'm doing here is generating the curated layer (which was generally handwritten) from typespec and patching in the REST translation, so I still need these methods for now. I'm happy to discuss in general what we are trying to do with generating models even for the curated layer

File: sdk/ml/azure-ai-ml/azure/ai/ml/operations/_marketplace_subscription_operations.py [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+        self,
+        operation_scope: OperationScope,
+        operation_config: OperationConfig,
+        service_client: ServiceClient202401Preview,
+    ):
+        super().__init__(operation_scope, operation_config)
+        self._service_client = service_client.marketplace_subscriptions
+
+    @experimental
+    def begin_create_or_update(
```
[the remark this answers is not in the document]
Comment (needuv) [reply sdk/ml/azure-ai-ml/azure/ai/ml/operations/_marketplace_subscription_operations.py:31]: added

File: sdk/ml/azure-ai-ml/azure/ai/ml/entities/_endpoint/serverless_endpoint.py (hunk 3 above)
Comment (needuv) [reply sdk/ml/azure-ai-ml/azure/ai/ml/entities/_endpoint/serverless_endpoint.py]: removed

File: sdk/ml/azure-ai-ml/azure/ai/ml/operations/_serverless_endpoint_operations.py [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+        self._marketplace_subscriptions = service_client.marketplace_subscriptions
+        self._all_operations = all_operations
+
+    def _get_workspace_location(self) -> str:
+        return str(
+            self._all_operations.all_operations[AzureMLResourceType.WORKSPACE].get(self._workspace_name).location
+        )
+
+    @experimental
+    def begin_create_or_update(self, endpoint: ServerlessEndpoint) -> LROPoller[ServerlessEndpoint]:
```
[the remark this answers is not in the document]
Comment (needuv) [reply sdk/ml/azure-ai-ml/azure/ai/ml/operations/_serverless_endpoint_operations.py]: added

File: sdk/ml/azure-ai-ml/azure/ai/ml/entities/_autogen_entities/models/_patch.py (hunk 2 above)
Comment (needuv) [reply sdk/ml/azure-ai-ml/azure/ai/ml/entities/_autogen_entities/models/_patch.py:83]: We want to figure out a way to streamline adding new models to the SDK, and this is a good opportunity for us to figure out what to do with the existing models as well. Will set up time with you to discuss

File: sdk/ml/azure-ai-ml/azure/ai/ml/_ml_client.py [hunk 6]
```diff
@@ -670,9 +675,22 @@ def __init__(
...
+        self._serverless_endpoints = ServerlessEndpointOperations(
+            self._operation_scope,
+            self._operation_config,
+            self._service_client_01_2024_preview,
+            self._operation_container,
+        )
+        self._marketplace_subscriptions = MarketplaceSubscriptionOperations(
+            self._operation_scope,
+            self._operation_config,
+            self._service_client_01_2024_preview,
```
Comment (MilesHolland) [inline sdk/ml/azure-ai-ml/azure/ai/ml/_ml_client.py:687]: Quick check: These are fine with the 01 API restclient?

Comment (needuv) [reply sdk/ml/azure-ai-ml/azure/ai/ml/_ml_client.py:687]: yup, they are

File: sdk/ml/azure-ai-ml/azure/ai/ml/entities/_autogen_entities/models/_models.py (hunk 1 above)
Comment (needuv) [reply sdk/ml/azure-ai-ml/azure/ai/ml/entities/_autogen_entities/models/_models.py]: only exposing necessary models for now

**Review by annatisch — APPROVED** [review]

File: sdk/ml/azure-ai-ml/azure/ai/ml/entities/_autogen_entities/models/_patch.py [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+            headers=obj.properties.inference_endpoint.headers,
+        )
+
+    def as_dict(self, *, exclude_readonly: bool = False) -> Dict[str, Any]:
+        d = super().as_dict(exclude_readonly=exclude_readonly)
+        d["system_data"] = json.loads(json.dumps(self.system_data._to_dict()))
+        return d
+
+
+ServerlessEndpoint.__doc__ += _ServerlessEndpoint.__doc__.strip() + """
```
Comment (luigiw) [inline sdk/ml/azure-ai-ml/azure/ai/ml/entities/_autogen_entities/models/_patch.py:123]: This is very smart! Thank you. Do we need to add doc string for other top level properties like `model_id` and `scoring_uri`?

**Review by luigiw — APPROVED** [review]

**Pull request merged**